src/spinwave.rs

```diff
@@ -128,53 +245,218 @@ fn spinwave_single_q(
+    );
+    let phase_factors_matrix = phase_factors.clone() * phase_factors.adjoint();
 
-    let n_sites = B.nrows();
+    let coefficients = component_mul(&(2. * spin_coefficients), &phase_factors_matrix);
 
-    let mut hamiltonian = Mat::<C64>::zeros(2 * n_sites, 2 * n_sites);
+    // We store S' as a 3x3 matrix of matrices indexed by alpha, beta
+    // where each element is an 2 * n_sites x 2 * n_sites matrix
+    // later on we will sum over the diagonal elements to get S'^alpha,beta(k, omega)
+    let mut sab_blocks = Mat::<Mat<C64>>::full(3, 3, Mat::<C64>::new());
 
-    // `submatrix_mut` takes a writable view of part of the matrix, and `copy_from` copies
-    // its argument into the matrix. So this is copying each of our matrices into sections
-    // of the `hamiltonian` matrix
-    // copy A - C into upper-left quarter of matrix
-    hamiltonian.submatrix_mut(0, 0, n_sites, n_sites).copy_from(A_minus_C);
-    // copy B into upper-right
-    hamiltonian.submatrix_mut(n_sites, 0, n_sites, n_sites).copy_from(B.as_ref());
-    // copy B* into bottom-left
-    hamiltonian.submatrix_mut(0, n_sites, n_sites, n_sites).copy_from(B.adjoint());
-    // copy A* - C into bottom-right
-    hamiltonian.submatrix_mut(n_sites, n_sites, n_sites, n_sites).copy_from(A_conj_minus_C);
+    // note one can show:
+    // Y*[alpha, beta] = Y[beta, alpha]
+    // Z*[alpha, beta] = V[beta, alpha]
+    // V*[alpha, beta] = Z[beta, alpha]
+    // W*[alpha, beta] = W[beta, alpha]
+    // thus we only need to calculate one triangle and can fill in the rest by conjugation
+    for alpha in 0..3 {
+        let z_alphas = Col::<C64>::from_iter(z.iter().map(|zi| zi[alpha]));
+        for beta in 0..=alpha {
+            let z_betas = Col::<C64>::from_iter(z.iter().map(|zi| zi[beta]));
 
-    hamiltonian
+            // construct the four blocks V, W, Y, Z;
+            // note that before we include the phase factors,
+            // V is conj(Z) and W is conj(Y)
+            let mut Yab = z_alphas.clone() * z_betas.adjoint();
+            let mut Zab = z_alphas.clone() * z_betas.transpose();
+            let mut Vab = Zab.conjugate().to_owned();
+            let mut Wab = Yab.conjugate().to_owned();
+
+            Yab = component_mul(&Yab, &coefficients);
+            Zab = component_mul(&Zab, &coefficients);
+            Vab = component_mul(&Vab, &coefficients);
+            Wab = component_mul(&Wab, &coefficients);
+
+            sab_blocks[(alpha, beta)] = block_matrix(&Yab, &Zab, &Vab, &Wab);
+
+            if beta < alpha {
+                let Yba = Yab.adjoint().to_owned();
+                let Zba = Vab.adjoint().to_owned();
+                let Vba = Zab.adjoint().to_owned();
+                let Wba = Wab.adjoint().to_owned();
+
+                sab_blocks[(beta, alpha)] = block_matrix(&Yba, &Zba, &Vba, &Wba);
+            }
+        }
+    }
+
+    let eigendecomp = (sqrt_hamiltonian.adjoint() * shc)
```

Comment:
The test failure on the `antiferro_chain` test is due to eigenvectors of degenerate energy levels not being unique or orthogonal; hence numerical differences between the Rust and Python code will give different `sab` values for the degenerate modes. We can add a small term to the diagonal of the matrix to be diagonalised to lift (numerically) the degeneracies to force the algorithm to make all eigenvectors orthogonal, by changing the above line to:

```suggestion
    // Adds a small delta to diagonal to ensure we don't have an exact degeneracies
    let diag_delta = Mat::<C64>::from_fn(shc.nrows(), shc.ncols(),
        |i,j| if i == j { C64::from((i as f64)*1e-12) } else { C64::ZERO } );
    let eigendecomp = ((sqrt_hamiltonian.adjoint() * shc) + diag_delta)
```

I found that a delta of `1e-12` was the smallest I could use and still get the test to pass, smaller values gave non-orthogonal eigenvectors and caused the test to fail - even then the test would fail unless I also added a relative tolerance of `1e-3`, or changed the absolute tolerance to `2e-3`.